sw/device/silicon_creator/lib/rescue/dfu.c

```diff
@@ -0,0 +1,274 @@
+
+#include "sw/device/silicon_creator/lib/rescue/dfu.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/drivers/usb.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+typedef struct rescue_mode_properties {
+  uint32_t mode;
+  bool dnload;
+  bool upload;
+} rescue_mode_properties_t;
+
+static const rescue_mode_properties_t mode_by_altsetting[] = {
+    {kRescueModeFirmware, true, false},
+    {kRescueModeFirmwareSlotB, true, false},
+    {kRescueModeOpenTitanID, false, true},
+    {kRescueModeBootLog, false, true},
+    {kRescueModeBootSvcRsp, true, true},
+    {kRescueModeOwnerPage0, true, true},
+    //{ kRescueModeOwnerPage1, false, true },
+};
+
+static rom_error_t dfu_control(dfu_ctx_t *ctx, usb_setup_data_t *setup) {
+  rom_error_t result = kErrorOk;
+  if (setup->request > kDfuReqAbort) {
+    ctx->dfu_state = kDfuStateError;
+    ctx->dfu_error = kDfuErrUnknown;
+    return kErrorUsbBadSetup;
+  }
+
+  dfu_state_transition_t *tr = &dfu_state_table[setup->request][ctx->dfu_state];
+
+  // Carry out the action for this state transition.
+  switch (tr->action) {
+    case kDfuActionNone:
+      // No action: move to the next state and ack the transaction.
+      ctx->dfu_state = tr->next[0];
+      dfu_transport_data(ctx, kUsbDirIn, NULL, 0, 0);
+      break;
+    case kDfuActionStall:
+      // Stall: move to the next state and stall the transport.
+      ctx->dfu_state = tr->next[0];
+      result = kErrorUsbBadSetup;
+      break;
+    case kDfuActionDataXfer:
+      // Check the length and download/upload.
+      ctx->dfu_state = tr->next[setup->length == 0 ? 0 : 1];
+      // Length is good and the request is either upload/download.
+      if (setup->length <= sizeof(ctx->state.data) &&
+          (setup->request == kDfuReqDnLoad ||
+           setup->request == kDfuReqUpLoad)) {
+        if (setup->request == kDfuReqDnLoad) {
+          // If its a download (transfer to opentitan), perform the transfer
+          // into the rescue state buffer. For a zero-length request, simply
+          // ACK with a zero-length packet.
+          dfu_transport_data(ctx, setup->length == 0 ? kUsbDirIn : kUsbDirOut,
+                             ctx->state.data, setup->length, 0);
+        } else {
+          // If its an upload (transfer to the host), perform the transfer
+          // from the rescue state buffer. A prior `SetInterface` command will
+          // have staged the requested content into the buffer.
+          size_t length = (size_t)MIN(ctx->state.staged_len, setup->length);
+          usb_transfer_flags_t flags =
+              length < setup->length && length % 64 == 0
+                  ? kUsbTransferFlagsShortIn
+                  : 0;
+          dfu_transport_data(ctx, kUsbDirIn, ctx->state.data, length, flags);
```

Comment:
Don't we need to track the current offset for upload blocks? Do we expect all uploaded data to be shorter than a max sized packet? (Am I missing where this is done?)